layers/org/funcs.el

```diff
@@ -0,0 +1,89 @@
+    (save-excursion
+      (save-match-data
+
+        ;; Find TOC
+        (let ((last-pos 0)
+              (toc-line-count 0))
+          (progn (beginning-of-buffer)
+                 (while (search-forward-regexp toc-line-rexp nil t)
+                   (incf toc-line-count))
+                 (unless (= toc-line-count 1)
+                   (error (format (concat "Found %d lines annotated with TOC. "
+                                          "Should be exactly 1.") toc-line-count)))))
+
+        (let ((line-after-toc-head-beg-pos (progn
+                                             (beginning-of-buffer)
+                                             (search-forward-regexp toc-line-rexp)
+                                             (line-beginning-position 2)))
+              headings
+              toc-items-and-levels)
+
+          (progn  (goto-char line-after-toc-head-beg-pos)
+                  ;; Remove TOC items
+                  (while (looking-at-p toc-item-rexp) (kill-whole-line))
+                  ;; Find all headings
+                  (while (search-forward-regexp org-heading-rexp nil t)
+                    (push (buffer-substring-no-properties
+                           (line-beginning-position)
+                           (line-end-position))
+                          headings))
+                  ;; Calculate headings levels
+                  (mapcar (lambda (heading)
+                            (let ((heading-lvl (- (length heading)
+                                                  (length (replace-regexp-in-string
+                                                           "^[\*]+"
+                                                           ""
+                                                           heading))))
+                                  ;; List of filters to be applied to the headings
+                                  (filters `((lambda (h-ing)
+                                               (progn (mapcar (lambda (rm-char)
+                                                                (setq h-ing (remove rm-char h-ing)))
+                                                              github-link-rm-chars)
+                                                      h-ing))
+                                             ,(apply-partially 'replace-regexp-in-string "^[\s]*" "")
+                                             ,(apply-partially 'replace-regexp-in-string "[\s]*$" "")
+                                             ,(apply-partially 'replace-regexp-in-string " " "-")
+                                             downcase)))
+
+                              (progn
+                                ;; Apply filters
+                                (mapcar (lambda (filter) (setq heading (funcall filter heading))) filters)
+                                (push (cons heading heading-lvl) toc-items-and-levels))))
+                          (reverse headings))
```

Comment:
70-71
![](http://img0.joyreactor.cc/pics/post/full/%D0%9C%D0%B0%D1%80%D0%BA%D0%B5%D1%80-%D0%9D%D0%B8%D1%89%D0%B5%D0%B1%D1%80%D0%BE%D0%B4-%D0%A1%D0%B4%D0%B5%D0%BB%D0%B0%D0%B9-%D0%A1%D0%B0%D0%BC-Proble-Genius--%D0%9D%D0%B5%D1%82-%D0%B4%D0%B5%D0%BD%D0%B5%D0%B3-%D0%BD%D0%B0-Fender-%D0%BF%D0%B5%D1%81%D0%BE%D1%87%D0%BD%D0%B8%D1%86%D0%B0-%D1%83%D0%B4%D0%B0%D0%BB%D1%91%D0%BD%D0%BD%D0%BE%D0%B5-291180.png)